walk/cache.go

```diff
@@ -71,3 +71,35 @@ func (c *cache) getLoaded(rel string) (dirInfo, error) {
 	ce := e.(*cacheEntry)
 	return ce.info, ce.err
 }
+
+var globalCache *cache
+
+func setGlobalCache(c *cache) func() {
+	if globalCache != nil {
+		panic("globalCache already set")
+	}
+	globalCache = c
+	return func() { globalCache = nil }
+}
+
+// GetDirInfo returns the list of files and subdirectories contained in a
+// directory named by rel. It also returns the parsed build file or nil if
+// none was present. rel is a slash-separated path, relative to the repository
+// root directory or "" for the root directory itself. The returned values
+// must not be modified.
+//
+// GetDirInfo does not directly perform any I/O. Instead, it provides access
+// to an internal cache to an internal cache that Walk and Walk2 use to speed
+// up directory operations. For this reason, GetDirInfo must not be called
+// on a directory Gazelle has not visited.
+//
+// In general, language extensions should prefer to use the RegularFiles,
+// Subdirs, and File fields of language.GenerateArgs. This function returns
```

Comment:
@jayconrod note that the fields in `DirInfo` are quite different then those in `GenerateArgs` when `generation_mode` is being used.

Was this intentional?

Reply:
FWIW I've always accessed this info via [a patch](https://github.com/aspect-build/aspect-cli/blob/dd3563abbc5b2f2844b80e789bd4e7049dfb1e23/patches/bazelbuild_bazel-gazelle_aspect-fs-direntry.patch) for what I think is the exact same scenario as yours: [expanding globs in `Imports()`](https://github.com/aspect-build/aspect-cli/blob/dd3563abbc5b2f2844b80e789bd4e7049dfb1e23/gazelle/js/resolve.go#L72).

That glob expansion needs to support globbing into subdirectories though...

Reply:
I think it can also panic due to race conditions though. IIUC if you invoke `GetDirInfo` faster then gazelle is walking those directories in the background...

I don't think `GetDirInfo` and a custom glob implementation in `gazelle_cc` is the ideal way of solving this. Maybe we can at least have a common implementation in gazelle instead of exposing utils like `GetDirInfo` exposing gazelle internals (in a way that is not ideal for our use case anyway)?

Reply:
`GetDirInfo` should not panic as long as its contract is satisfied. That contract is rather complicated though, so if you see a panic, please open an issue and we can fix it if it's a bug or clarify the docs.

Reply:
Won't it panic if you try fetching a directory not loaded yet? Such as a subdirectory...? I'm speculating here so forgive me if I'm wrong :|